src/docs/testing/best-practices.md

```diff
@@ -0,0 +1,174 @@
+---
+title: Performance best practices
+short-title: Best practices
+description: How to ensure that your Flutter app is performant.
+---
+
+Generally Flutter applications are high performance by default,
+so you only need to avoid common pitfalls to get excellent performance
+instead of needing to micro-optimize with complicated profiling tools.
+These best recommendations will help you write the most performant
+Flutter app possible.
+
+## Best practices
+
+How do you design a Flutter app to most efficiently render your scenes?
+In particular, how do you ensure that the painting code generated by the
+framework is as efficient as possible? Here are a few things to consider
+when designing your app:
+
+### Call setState thoughtfully
+
+* Avoid repetitive and costly work in `build()` methods since `build()`
+  can be invoked frequently when ancestor Widgets rebuild.
+* Localize the `setState()` call to the part of the subtree whose UI
+  actually needs to change. Avoid calling calling setState() high up in
+  the tree if the change is contained to a small part of the tree.
+* Re-use the same child widget instance when building stops that
+  child subtree from re-building when the parent rebuilds.
+  [PENDING: rephrase]
+
+Also see:
+
+* [Performance considerations](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/StatefulWidget-class.html#performance-considerations),
+  part of the
+  [StatefulWidget](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/StatefulWidget-class.html)
+  API doc
+* [Improve your Flutter app performance: split your widgets
+  (with MVP)](https://medium.com/flutter-community/improve-your-flutter-app-performance-split-your-widgets-935f97e93f7d)
+  by Simon Lightfoot
+
+### Apply effects only when needed
+
+Use effects carefully, as they can be expensive. Some of them invoke
+`saveLayer()` behind the scenes, which can be an expensive operation.
+
+{{site.alert.note}}
+**Why is savelayer expensive?**<br>
+Calling `saveLayer()` allocates an offscreen buffer. Drawing content
+into the offscreen buffer may trigger render target switches that
+are particularly slow in older GPUs.
+{{site.alert.end}}
+
+Some general rules when applying specific effects:
+
+* Use the [**Opacity**]({{site.api}}/flutter/widgets/Opacity-class.html)
+  widget only when necessary, but if you must use it, add a
+  RepaintBoundary right below Opacity in the widget tree to prevent
+  unnecessary calls to `saveLayer()`.  See [Transparent
+  image]({{site.api}}/flutter/widgets/Opacity-class.html#transparent-image)
+  in the Opacity API page for an example of applying opacity directly
+  to an image, which is faster than using the Opacity widget.
+* **Clipping** doesn’t call `saveLayer()` (unless explicitly requested with
+  `Clip.antiAliasWithSaveLayer()`) so these operations aren’t as expensive
```

Comment:
`Clip.antiAliasWithSaveLayer()` -> `Clip.antiAliasWithSaveLayer` since `Clip.antiAliasWithSaveLayer` is an enum, not a function.